File: java/src/main/java/benchmark/model/v2/Cells.java
package benchmark.model.v2;

import benchmark.model.Fixture;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/** Resolve run-config cells via scripts/resolve_run_config.py (same as Go/Python). */
public final class Cells {
  private static final ObjectMapper MAPPER = new ObjectMapper();

  public record Cell(
      String typeId,
      Map<String, Object> typeConfig,
      String typeConfigHash,
      int dataTypeInstanceCount) {}

  public record ResolvedRun(List<Cell> cells, List<String> ioModes, long seed) {}

  public record WorkItem(Fixture fixture, int instanceCount, String typeConfigHash) {}

  private Cells() {}

  public static Path repoRoot() {
    Path dir = Path.of(System.getProperty("user.dir")).toAbsolutePath();
    for (Path p = dir; p != null; p = p.getParent()) {
      if (Files.isRegularFile(p.resolve("config/benchmark_config.yaml"))) {
        return p;
      }
    }
    return dir;
  }

  public static ResolvedRun loadResolved(String runConfigPath, long seed) throws Exception {
    Path root = repoRoot();
    Path script = root.resolve("scripts/resolve_run_config.py");
    if (runConfigPath == null || runConfigPath.isBlank()) {
      runConfigPath = root.resolve("config/library/default.yaml").toString();
    } else {
      Path candidate = root.resolve(runConfigPath);
      if (Files.isRegularFile(candidate)) {
        runConfigPath = candidate.toString();
      }
    }
    ProcessBuilder pb =
        new ProcessBuilder(
            "python3",
            script.toString(),
            runConfigPath,
            "--seed",
            Long.toString(seed));
    pb.directory(root.toFile());
    Map<String, String> env = pb.environment();
    String analysisSrc = root.resolve("analysis/src").toString();
    String prev = env.getOrDefault("PYTHONPATH", "");
    env.put("PYTHONPATH", prev.isEmpty() ? analysisSrc : analysisSrc + ":" + prev);
    pb.redirectErrorStream(true);
    Process proc = pb.start();
    StringBuilder out = new StringBuilder();
    try (BufferedReader br =
        new BufferedReader(new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8))) {
      String line;
      while ((line = br.readLine()) != null) out.append(line).append('\n');
    }
    int code = proc.waitFor();
    if (code != 0) {
      throw new IllegalStateException("resolve_run_config failed (" + code + "): " + out);
    }
    JsonNode rootNode = MAPPER.readTree(out.toString());
    List<Cell> cells = new ArrayList<>();
    for (JsonNode c : rootNode.path("cells")) {
      Map<String, Object> cfg = new HashMap<>();
      JsonNode tc = c.path("type_config");
      if (tc.isObject()) {
        Iterator<Map.Entry<String, JsonNode>> it = tc.fields();
        while (it.hasNext()) {
          Map.Entry<String, JsonNode> e = it.next();
          cfg.put(e.getKey(), MAPPER.convertValue(e.getValue(), Object.class));
        }
      }
      cells.add(
          new Cell(
              c.path("type_id").asText(),
              cfg,
              c.path("type_config_hash").asText(""),
              c.path("data_type_instance_count").asInt(1)));
    }
    List<String> modes = new ArrayList<>();
    for (JsonNode m : rootNode.path("execution").path("io_modes")) {
      modes.add(m.asText());
    }
    long resolvedSeed = seed;
    if (rootNode.hasNonNull("seed")) {
      resolvedSeed = rootNode.get("seed").asLong(seed);
    }
    return new ResolvedRun(cells, modes, resolvedSeed);
  }

  public static WorkItem fixtureFromCell(Cell cell, long seed) {
    int n = Math.max(1, cell.dataTypeInstanceCount());
    List<Object> insts = Generators.instances(cell.typeId(), cell.typeConfig(), seed, n);
    Object value;
    if (n == 1) {
      value = insts.get(0);
    } else {
      value =
          switch (cell.typeId()) {
            case "message" -> {
              List<Message> list = new ArrayList<>(n);
              for (Object o : insts) list.add((Message) o);
              yield list;
            }
            case "document" -> {
              List<Document> list = new ArrayList<>(n);
              for (Object o : insts) list.add((Document) o);
              yield list;
            }
            case "telemetry" -> {
              List<Telemetry> list = new ArrayList<>(n);
              for (Object o : insts) list.add((Telemetry) o);
              yield list;
            }
            case "strings" -> {
              List<Strings> list = new ArrayList<>(n);
              for (Object o : insts) list.add((Strings) o);
              yield list;
            }
            case "event" -> {
              List<Event> list = new ArrayList<>(n);
              for (Object o : insts) list.add((Event) o);
              yield list;
            }
            default -> insts;
          };
    }
    return new WorkItem(new Fixture(cell.typeId(), value), n, cell.typeConfigHash());
  }
}

File: kotlin/src/main/kotlin/benchmark/model/v2/Cells.kt
package benchmark.model.v2

import benchmark.model.Fixture
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

/** Resolve run-config cells via scripts/resolve_run_config.py (same as Java/Go/Python). */
object Cells {
    private val mapper = ObjectMapper()

    data class Cell(
        val typeId: String,
        val typeConfig: Map<String, Any?>,
        val typeConfigHash: String,
        val dataTypeInstanceCount: Int,
    )

    data class ResolvedRun(val cells: List<Cell>, val ioModes: List<String>, val seed: Long)

    data class WorkItem(val fixture: Fixture, val instanceCount: Int, val typeConfigHash: String)

    fun repoRoot(): Path {
        var dir = Path.of(System.getProperty("user.dir")).toAbsolutePath()
        var p: Path? = dir
        while (p != null) {
            if (Files.isRegularFile(p.resolve("config/benchmark_config.yaml"))) {
                return p
            }
            p = p.parent
        }
        return dir
    }

    fun loadResolved(runConfigPath: String?, seed: Long): ResolvedRun {
        val root = repoRoot()
        val script = root.resolve("scripts/resolve_run_config.py")
        var cfgPath = runConfigPath
        if (cfgPath.isNullOrBlank()) {
            cfgPath = root.resolve("config/library/default.yaml").toString()
        } else {
            val candidate = root.resolve(cfgPath)
            if (Files.isRegularFile(candidate)) {
                cfgPath = candidate.toString()
            }
        }
        val pb = ProcessBuilder(
            "python3",
            script.toString(),
            cfgPath,
            "--seed",
            seed.toString(),
        )
        pb.directory(root.toFile())
        val env = pb.environment()
        val analysisSrc = root.resolve("analysis/src").toString()
        val prev = env.getOrDefault("PYTHONPATH", "")
        env["PYTHONPATH"] = if (prev.isEmpty()) analysisSrc else "$analysisSrc:$prev"
        pb.redirectErrorStream(true)
        val proc = pb.start()
        val out = String(proc.inputStream.readAllBytes(), StandardCharsets.UTF_8)
        val code = proc.waitFor()
        if (code != 0) {
            throw IllegalStateException("resolve_run_config failed ($code): $out")
        }
        val rootNode = mapper.readTree(out)
        val cells = mutableListOf<Cell>()
        for (c in rootNode.path("cells")) {
            val cfg = mutableMapOf<String, Any?>()
            val tc = c.path("type_config")
            if (tc.isObject) {
                val it = tc.fields()
                while (it.hasNext()) {
                    val e = it.next()
                    cfg[e.key] = mapper.convertValue(e.value, Any::class.java)
                }
            }
            cells.add(
                Cell(
                    c.path("type_id").asText(),
                    cfg,
                    c.path("type_config_hash").asText(""),
                    c.path("data_type_instance_count").asInt(1),
                ),
            )
        }
        val modes = mutableListOf<String>()
        for (m in rootNode.path("execution").path("io_modes")) {
            modes.add(m.asText())
        }
        var resolvedSeed = seed
        if (rootNode.hasNonNull("seed")) {
            resolvedSeed = rootNode.get("seed").asLong(seed)
        }
        return ResolvedRun(cells, modes, resolvedSeed)
    }

    fun fixtureFromCell(cell: Cell, seed: Long): WorkItem {
        val n = maxOf(1, cell.dataTypeInstanceCount)
        val insts = Generators.instances(cell.typeId, cell.typeConfig, seed, n)
        val value: Any =
            if (n == 1) {
                insts[0]
            } else {
                when (cell.typeId) {
                    "message" -> insts.map { it as Message }.toMutableList()
                    "document" -> insts.map { it as Document }.toMutableList()
                    "telemetry" -> insts.map { it as Telemetry }.toMutableList()
                    "strings" -> insts.map { it as Strings }.toMutableList()
                    "event" -> insts.map { it as Event }.toMutableList()
                    else -> insts
                }
            }
        return WorkItem(Fixture(cell.typeId, value), n, cell.typeConfigHash)
    }
}
